Add input lvgl binding

faxe1008 opened the pull request:
Add an `zephyr,input-lvgl` device which can handle the `input-events` generated by the various touch input drivers directly. Additionally this allows for usage of other LVGL `indev` types. At the moment only additional properties are added for the pointer type to have parity with the existing Kconfig options of the lvgl module.

The only other device for which it might make sense to have properties would be the `button` `indev` type which can be assigned coordinates on the screen. I am unsure if this use case warrants a static property, because you might want to set those depending on the screen you are currently on.

This addresses the draft of [https://github.com/zephyrproject-rtos/lvgl/pull/43](https://github.com/zephyrproject-rtos/lvgl/pull/43). However there is one last issue: The implementation depends on the `struct lvgl_disp_data` from the `lvgl_display.h` header. I don't think there is a (clean) way to avoid also touching the module in any way.

faxe1008 committed:
modules: lvgl: add zephyr,lvgl-pointer-input pseudo device

Add the scaffolding to create input lvgl pseudo devices which route zephyr
input_event to their lvgl `indev` equivalent. As a first cut also add a
`zephyr,lvgl-pointer-input compatible which can be a drop-in replacement
for the existing kscan solution.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
boards: posix: native_posix: Rename input-sdl-touch node

Rename the input-sdl-touch node to be able to have phandle references to
it.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
samples: subsys: display: lvgl: Use input-lvgl for touch input

Changes all board specific configurations to use the new
zephyr,lvgl-pointer-input compatible. The kscan_input node in the boards
remains to have downstream backwards compatibility.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

gmarull commented on dts/bindings/input/zephyr,input-lvgl.yaml:
```diff
@@ -0,0 +1,58 @@
+# Copyright 2023 Fabian Blatz <fabianblatz@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  LVGL indev pseudo-device
+
+  Listens for input events and routes the
+  lv_indev_data_t to the underlying lv_indev_t managed by LVGL.
+
+  Needs to be associated to a specific device to listen for events
+  from that device. Example configuration:
+
+  pointer {
+          compatible = "zephyr,input-lvgl";
+          input = <&input_sdl_touch>;
+          type = "pointer";
+  };
+
+
+compatible: "zephyr,input-lvgl"
+
+properties:
+  input:
+    required: true
+    type: phandle
+    description: |
+      Input device phandle, if not specified listen for input from all devices.
+
+  type:
+    type: string
+    required: true
+    enum:
+      - "pointer"
+      - "keypad"
+      - "button"
+      - "encoder"
+    description: LVGL indev type.
+  
+  msgq-size:
```
I'm a bit worried about the proliferation of "software devices" in devicetree, in particular here we are using devicetree to define software-specific things like a message queue size.

I think this needs discussion, and potentially alternative solutions. This was not needed by e.g. kscan, so why is this added for input?

fabiobaltieri replied:
msgq-size should probably remain an option, but the various swap-... I think belong to here since they are a hardware characteristic, plus you could imagine supporting multiple devices with different setups.

The other thing that should go here is the mapping of axis and button codes, which matches the ones emitted by a `gpio-keys` for example, so I think it makes sense for those too.

faxe1008 replied:
> devicetree to define software-specific things like a message queue size.

Agreed, I will move the `msgq-size` back to being a Kconfig option. 

> This was not needed by e.g. kscan, so why is this added for input?

Yeah, the way LVGL expects to be fed input event needs a message queue of some sorts. Previously this was a Kconfig option in the LVGL module.

> The other thing that should go here is the mapping of axis and button codes

For the axis I am not sure. Lots of LVGL code either makes use of the `enc_diff` value or it expects that the encoder really only emits `LV_KEY_LEFT` or `LV_KEY_RIGHT`: [https://docs.lvgl.io/master/overview/indev.html](https://docs.lvgl.io/master/overview/indev.html), e.g. [lv_arc](https://github.com/lvgl/lvgl/blob/v8.3.7/src/widgets/lv_arc.c#L554) . I suppose I'd need to set the `lv_indev_data_t->code` depending on the `input_event->value` sign what do you think?

fabiobaltieri replied:
> For the axis I am not sure. Lots of LVGL code either makes use of the `enc_diff` value or it expects that the encoder really only emits `LV_KEY_LEFT` or `LV_KEY_RIGHT`: https://docs.lvgl.io/master/overview/indev.html, e.g. [lv_arc](https://github.com/lvgl/lvgl/blob/v8.3.7/src/widgets/lv_arc.c#L554) . I suppose I'd need to set the `lv_indev_data_t->code` depending on the `input_event->value` sign what do you think?

Right but you need to decide what event code to filter on, for example `INPUT_REL_Z`. For touchscreen all drivers send X/Y but for wheels and buttons it's up to the driver and you want to match it so we need to tell the LVGL input code somehow, and I think it makes sense to have it in the dts file.

faxe1008 replied:
Ah ok, so this was meant for filtering purposes. I thought you meant adding the axis data somehow into `lv_indev_data_t`.

fabiobaltieri replied:
Yeah, or any other configuration that depends on the hardware setup, I imagine stuff like `btn_points` data could go there.

jfischer-no reviewed (changes requested):
LVGL is not a part of Zephyr and not a Zephyr project, therefore something like [subsys/input/input_lvgl.c](https://github.com/zephyrproject-rtos/zephyr/pull/60867/commits/6de246add7f4e6fe695d747ef71acd8287e9a2a3#diff-98dd9f8dd7f65377341b1c8fc5ce9ec1fe3ffd8d19e0667fe68e149913bed967) cannot be in subsystem/input/.

fabiobaltieri commented:
> I think @fabiobaltieri opened this issue to clarify this #60868. In the end I am also fine with moving the changes into LVGL module.

Yeah there's few modules with module specific code in https://github.com/zephyrproject-rtos/zephyr/tree/main/modules, we should settle on something. We can have that discussion on the issue and have the (initial) review for this one here.

fabiobaltieri commented on subsys/input/input_lvgl.c:
```diff
@@ -0,0 +1,251 @@
+#define INPUT_LVGL_MSGQ_SIZE(inst) (DT_INST_PROP(inst, msgq_size))
+
+#define ASSERT_PROPS(inst)                                                                         \
+	BUILD_ASSERT(!DT_INST_PROP_OR(inst, swap_xy, 0) ||                                         \
+			     DT_INST_ENUM_IDX(inst, type) == BINDING_TYPE_POINTER,                 \
+		     "Property swap-xy only supported for pointer type.");                         \
+	BUILD_ASSERT(!DT_INST_PROP_OR(inst, invert_x, 0) ||                                        \
+			     DT_INST_ENUM_IDX(inst, type) == BINDING_TYPE_POINTER,                 \
+		     "Property invert-x only supported for pointer type.");                        \
+	BUILD_ASSERT(!DT_INST_PROP_OR(inst, invert_y, 0) ||                                        \
+			     DT_INST_ENUM_IDX(inst, type) == BINDING_TYPE_POINTER,                 \
+		     "Property invert-y only supported for pointer type.")
```
I think you'd be better off with a binding per type (like a `-common.yaml` file and then an `-pointer.yaml`...), then you'd get these checks automatically from the binding. Same for the config and data structures, you could have a common and specific one, it'd make the whole thing more compact memory wise.

faxe1008 replied:
I implemented your suggested changes. You were correct, this feels cleaner now :+1:

faxe1008 committed:
input: add zephyr,lvgl-button-input device binding

Add a pseudo device which can be used to hook into gpio-keys input_events
and relay the events to a lv_indev.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

fabiobaltieri commented on dts/bindings/input/zephyr,input-lvgl-common.yaml:
```diff
@@ -0,0 +1,11 @@
+
+properties:
+  input:
+    required: true
```
let's keep this optional, would be useful for situations like having multiple button devices, or even just for testing and having the application generate events

fabiobaltieri commented on subsys/input/CMakeLists.txt:
```diff
@@ -6,3 +6,12 @@ zephyr_library_sources(input.c)
+
+if(CONFIG_INPUT_LVGL)
+
+zephyr_include_directories(${ZEPHYR_BASE}/subsys/input)
```
I don't think this is needed since the common file is include with quotes (`#include "input_lvgl_common.h"`) and the files that use it are in the same directory.

fabiobaltieri commented on subsys/input/Kconfig:
```diff
@@ -86,4 +86,32 @@ config INPUT_LONGPRESS
+
+if INPUT_LVGL_POINTER
+
+config INPUT_LVGL_POINTER_MSGQ_COUNT
+	int "Message queue size"
+	default 10
+	help
+	  Size of the pointer message queue buffering input events.
```
how about having a single config for the queue size for all input events? keep the config a bit simpler

faxe1008 replied:
Thought about that one for a bit, I wanted to have this per `indev` type because, a touch controller with interrupt generated `input_events` will want to have a bigger queue then a button, that you can most likely only press once within the lvgl refresh period. Tried saving some flash :^)

fabiobaltieri replied:
Sounds good.

fabiobaltieri commented on subsys/input/input_lvgl_common.c:
```diff
@@ -0,0 +1,48 @@
+
+#include "input_lvgl_common.h"
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(input_lvgl, CONFIG_INPUT_LOG_LEVEL);
```
This may use a log level for the module rather than the input one, again, post code move comment.

faxe1008 replied:
Changed to `LOG_MODULE_DECLARE(lvgl)`

fabiobaltieri commented on subsys/input/input_lvgl_common.c:
```diff
@@ -0,0 +1,48 @@
+	lv_indev_drv_init(&common_data->indev_drv);
+	common_data->indev_drv.type = indev_type;
+	common_data->indev_drv.read_cb = lvgl_input_read_cb;
+	common_data->indev = lv_indev_drv_register(&common_data->indev_drv);
+	common_data->indev_drv.user_data = common_data;
```
the sequence of these two looks funny... can they be swapped? would also be nice to have the `indev == NULL` check just next to the `lv_indev_drv_register` call

fabiobaltieri commented on samples/subsys/display/lvgl/src/main.c:
```diff
@@ -77,7 +77,8 @@ int main(void)
 
-	if (IS_ENABLED(CONFIG_LV_Z_POINTER_KSCAN)) {
+	if (IS_ENABLED(CONFIG_LV_Z_POINTER_KSCAN) ||
+	    IS_ENABLED(CONFIG_DT_HAS_ZEPHYR_INPUT_LVGL_POINTER_ENABLED)) {
```
use `INPUT_LVGL_POINTER` here, normally those `DT_HAS` should only be used for dependencies

fabiobaltieri commented on dts/bindings/input/zephyr,input-lvgl-button.yaml:
```diff
@@ -0,0 +1,47 @@
+  pointer {
+          compatible = "zephyr,input-lvgl-button";
+          input = <&buttons>;
+          input-codes = <INPUT_KEY_0>;
+          lvgl-key = <10>;
+          x-coordinate = <120>;
+          y-coordinate = <120>;
```
Oh that's interesting, so the button can either map to an `lvgl-key` or a coordinate right? How about splitting this in two, `zephyr,input-lvgl-button-key` and `zephyr,input-lvgl-button-coordinate`, then you would have an array of codes mapping to either an array or keys or an array of coordinates. That way you could do all keys to whatever combination of coordinate and (lvgl) keys in at most two definitions, and also saves you from having to do build check based validation.

faxe1008 replied:
From what I understood and also while testing it seems like it goes like this:
If the assigned `lv_indev_data_t->key` is `LV_KEY_ENTER (10)` then the press and release are relayed as a click at the coordinates of the button.
Otherwise there are keys with a special interaction: [https://docs.lvgl.io/latest/en/html/overview/indev.html](https://docs.lvgl.io/latest/en/html/overview/indev.html)

fabiobaltieri reviewed (commented):
Hey good stuff, love where this is going. Did a first pass, a bunch of stuff will move and make more sense once the files have been moved over to the main repo, it may be worth putting this on hold until those are moved over.

faxe1008 commented on subsys/input/input_lvgl_pointer.c:
```diff
@@ -0,0 +1,135 @@
+{
+	struct input_lvgl_common_data *data = dev->data;
+
+	return register_lvgl_driver(LV_INDEV_TYPE_POINTER, data);
```
Yeah, that generated qualifier discard warnings, thats why I went with the data, but true might be more future-proof.

faxe1008 commented on subsys/input/input_lvgl_pointer.c:
```diff
@@ -0,0 +1,135 @@
+		.invert_y = DT_INST_PROP(inst, invert_y),                                          \
+	};                                                                                         \
+	static struct input_lvgl_common_data input_lvgl_common_data##inst = {                      \
+		.event_msgq = &INPUT_LVGL_EVENT_MSGQ(inst, pointer)};                              \
```
Ah, did not know about that concern. Moved it.

faxe1008 commented:
@fabiobaltieri addressed all of your comments that would not be affected by the code move. Also noticed the problem with the  defconfigs for some of the boards in regards to the LVGL sample. Will tackle the move and the rest of this after my vacation :^)

fabiobaltieri commented on samples/subsys/display/lvgl/boards/stm32f746g_disco.overlay:
```diff
@@ -0,0 +1,16 @@
+		status = "okay";
+		compatible = "zephyr,input-lvgl-pointer";
+		input = <&ft5336>;
+	};
```
The two symbols should probably be exclusive, I don't see a good reason to mix them up, but also once this is in I think it's fair to remove all the `config KSCAN default y if LVGL` from the boards and just use `config INPUT default y if LVGL`.

faxe1008 replied:
I just double checked, boards or shields that default `LV_Z_POINTER_KSCAN` are:
```
mimxrt1050_evk
mimxrt1060_evk
mimxrt1064_evk
bl5340_dvk
rk055hdmipi4ma0
buydisplay_2_8_tft_touch_arduino
buydisplay_3_5_tft_touch_arduino
g1120b0mipi
adafruit_2_8_tft_touch_v2
```

The two physical boards that I enabled the new `zephyr,lvgl-pointer-input` for do not enable the symbol to be `y` by default, so we should be good here. Arguably in a second step the boards/shields above should be migrated.

fabiobaltieri replied:
Oh I see what you mean. Yeah fine to fix the board in a follow up PR.

fabiobaltieri commented:
> @fabiobaltieri addressed all of your comments that would not be affected by the code move. Also noticed the problem with the defconfigs for some of the boards in regards to the LVGL sample. Will tackle the move and the rest of this after my vacation :^)

Have a nice break!

gmarull commented on dts/bindings/input/zephyr,input-lvgl-pointer.yaml:
```diff
@@ -0,0 +1,36 @@
+          input = <&input_sdl_touch>;
+  };
+
+compatible: "zephyr,input-lvgl-pointer"
```
I'm a bit concerned about bindings specific to 3rd party modules. I think we need to avoid this.

gmarull replied:
@mbolivar-ampere @galak

mbolivar-ampere replied:
Yeah, this seems potentially borderline. Does it comply with the rules in https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#the-zephyr-prefix or is it violating them somehow?

fabiobaltieri replied:
How about using an `lvgl,` prefix for these? Would make them more compact as well, `lvgl,input-pointer`.

Anyway I think @gmarull was referring to the fact that the binding is for a module, not the name itself. I think it's fine in this case (or better, I don't see a better alternative) but yeah worth discussing.

gmarull replied:
Well, actually I see other precedents (openamp):

> You may use the zephyr, prefix when naming a devicetree compatible that is specific to Zephyr. One example is [zephyr,ipc-openamp-static-vrings](https://docs.zephyrproject.org/latest/build/dts/api/bindings/ipc/zephyr%2Cipc-openamp-static-vrings.html#std-dtcompatible-zephyr-ipc-openamp-static-vrings). In this case, it’s permitted but not required to add the zephyr, prefix to properties defined in the binding.

So no objections. I wonder though if something like this is LVGL specific or could be made generic for other libraries.

mbolivar-ampere replied:
I think an `lvgl` prefix is probably worse than `zephyr` -- we have reasonable expectation that the project is big enough that the zephyr prefix is "ours" at this point, kind of like how linux has sometimes at least historically used `linux` as a prefix. So we can hide our mistakes in our namespace 😉 

If this can be made generic, all the better, but that sounds like a separate question. Overall my conclusion is that using the zephyr prefix here is OK.

gmarull commented on subsys/input/input_lvgl_common.c:
```diff
@@ -0,0 +1,55 @@
+	common_data->indev_drv.type = indev_type;
+	common_data->indev_drv.read_cb = lvgl_input_read_cb;
+	common_data->indev_drv.user_data = dev;
+	common_data->indev = lv_indev_drv_register(&common_data->indev_drv);
```
I feel like we should have glue code (in modules/) calling Zephyr APIs and not the other way around.

fabiobaltieri replied:
yeah this whole stuff have to be moved in `modules/lvgl`

faxe1008 commented:
@fabiobaltieri I'll be working on rebasing this one. Whats is your opinion on preserving compatibility with all the `LV_Z_KSCAN_POINTER` options?

- Create another temporary pseudo device which uses the kscan one as input
- Leave `Kconfig.input` options and `lvgl.c` code as is.
- Separate out the code from `lvgl.c` and have it in its own `SYS_INIT` function in a `lvgl_kscan_pointer.c` file which is only selected if the `LV_Z_KSCAN_POINTER` option is selected. 

I'd prefer option 3, although this will cause some code duplication for the screen rotation handling.

faxe1008 committed:
modules: lvgl: Move lvgl kscan pointer handling to separate file

Moves lvgl pointer driver handling based on kscan to its own file, to ease
deletion when the usage of kscan for display touch input has been
deprecated.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 commented on subsys/input/input_lvgl_common.c:
```diff
@@ -0,0 +1,48 @@
+ * initialized. Also since this device relies on the input subsystem, ensure that this one is ready
+ * too.
+ */
+BUILD_ASSERT(CONFIG_APPLICATION_INIT_PRIORITY < CONFIG_INPUT_LVGL_INIT_PRIORITY);
```
Moved to `lvgl.c`

faxe1008 commented:
@fabiobaltieri I restructured the PR and moved the device drives into `modules/lvgl/input`. I also renamed the compatibles  from `zephyr,input-lvgl-X` to `zephyr,lvgl-X-input` since it rolls nicer of the tongue and also is consistent with the existing `LV_Z_POINTER_KSCAN` symbol.

I skimmed again through your previous comments. The only thing left would be the board defkonfigs and what to do about them.

fabiobaltieri commented on modules/lvgl/input/lvgl_pointer_input.c:
```diff
@@ -0,0 +1,133 @@
+	return register_lvgl_indev_driver(LV_INDEV_TYPE_POINTER, dev);
+}
+
+#define LVGL_POINTER_INPUT_DEFINE(_inst)                                                           \
```
Could you just use `_inst`? don't think there's any naming conflict in here

faxe1008 replied:
Is this a typo? Did you mean changing the macro parameter name from `_inst` to `inst`?

fabiobaltieri replied:
yeah sorry I mean use `inst`

fabiobaltieri commented on modules/lvgl/input/lvgl_common_input.h:
```diff
@@ -0,0 +1,52 @@
+		process_input_event(DEVICE_DT_INST_GET(_inst), evt);                               \
+	}                                                                                          \
+	INPUT_CALLBACK_DEFINE(LVGL_INPUT_DEVICE(_inst), lvgl_input_cb_##_inst);                    \
+	K_MSGQ_DEFINE(event_msgq_##_type##_inst, sizeof(lv_indev_data_t), _msgq_size, 4)
```
reuse `LVGL_INPUT_EVENT_MSGQ` here so you don't have the name format duplicated

faxe1008 replied:
Trying to use the macro in here yields a expansion error:
```
zephyr/kernel.h:4446:43: warning: implicit declaration of function ‘_k_fifo_buf_LVGL_INPUT_EVENT_MSGQ’ [-Wimplicit-function-declaration]
 4446 |                Z_MSGQ_INITIALIZER(q_name, _k_fifo_buf_##q_name, \
```
Not sure how to resolve this.

fabiobaltieri replied:
I see, I think it's because `K_MSGQ_DEFINE` adds another prefix, so this ends up being `_k_fifo_buf_LVGL_INPUT_EVENT_MSGQ(0, pointer)` and does not expand anymore (you can find it by building with `CONFIG_COMPILER_SAVE_TEMPS=y` and then checking what ends up in `build/modules/lvgl/CMakeFiles/modules__lvgl.dir/input/lvgl_pointer_input.c.i`)

Ok let's leave it like this then. The alternative would be to put the queue and buffer separately in the data structure and initialize it in runtime, see https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/can/can_loopback.c#L34-L35 for example but then yo'd have to initialize it for every type of input device.

faxe1008 replied:
I think I'd rather go without separating it, makes init code more complicated 👍

fabiobaltieri commented on modules/lvgl/input/lvgl_common_input.h:
```diff
@@ -0,0 +1,52 @@
+
+int register_lvgl_indev_driver(lv_indev_type_t indev_type, const struct device *dev);
+
+#define LVGL_INPUT_EVENT_MSGQ(_inst, _type) event_msgq_##_type##_inst
```
I would add some extra `_` between type and inst: `event_msgq_##_type##_##_inst` so it's coherent with other structures), also change the prefix to match the rest here, maybe `lvgl_input_msg...`. Just makes your life easier to find all correlated symbols if you have to.

fabiobaltieri commented on modules/lvgl/input/lvgl_common_input.h:
```diff
@@ -0,0 +1,52 @@
+int register_lvgl_indev_driver(lv_indev_type_t indev_type, const struct device *dev);
+
+#define LVGL_INPUT_EVENT_MSGQ(_inst, _type) event_msgq_##_type##_inst
+#define LVGL_INPUT_DEVICE(_inst)            DEVICE_DT_GET_OR_NULL(DT_INST_PHANDLE(_inst, input))
+
+#define LVGL_COORD_VALID(_coord) (_coord >= LV_COORD_MIN && _coord <= LV_COORD_MAX)
+#define LVGL_KEY_VALID(_key)     (_key >= 0 && _key <= 255)
```
Drop the _ prefix if it's not needed, same in the other macros.

fabiobaltieri commented on modules/lvgl/input/lvgl_common_input.h:
```diff
@@ -0,0 +1,52 @@
+#define LVGL_KEY_VALID(_key)     (_key >= 0 && _key <= 255)
+
+#define LVGL_INPUT_DECLARE(_inst, _type, _msgq_size)                                               \
+	static void process_input_event(const struct device *dev, struct input_event *evt);        \
```
is this forward declaration needed?

fabiobaltieri commented on modules/lvgl/input/lvgl_common_input.c:
```diff
@@ -0,0 +1,47 @@
+	data->continue_reading = k_msgq_num_used_get(cfg->event_msgq) > 0;
+}
+
+int register_lvgl_indev_driver(lv_indev_type_t indev_type, const struct device *dev)
```
use a common function prefix, `lvgl_input` or something

fabiobaltieri commented on modules/lvgl/input/lvgl_pointer_input.c:
```diff
@@ -0,0 +1,133 @@
+
+	switch (evt->code) {
+	case INPUT_ABS_X:
+		data->pending_event.point.x = evt->value;
```
Use a temporary variable for `data->pending_event.point`, should make everything way more compact.

fabiobaltieri commented on modules/lvgl/input/lvgl_pointer_input.c:
```diff
@@ -0,0 +1,133 @@
+
+	/* adjust coordinates */
+	if (cfg->swap_xy) {
+		lv_coord_t x;
```
tmp? x sounds like horizontal axis coordinate

fabiobaltieri commented on modules/lvgl/input/lvgl_pointer_input.c:
```diff
@@ -0,0 +1,133 @@
+
+	/* rotate touch point to match display rotation */
+	if (cap->current_orientation == DISPLAY_ORIENTATION_ROTATED_90) {
+		lv_coord_t x;
```
same

fabiobaltieri commented on modules/lvgl/input/lvgl_pointer_input.c:
```diff
@@ -0,0 +1,133 @@
+		data->pending_event.point.x = cap->x_resolution - data->pending_event.point.x;
+		data->pending_event.point.y = cap->y_resolution - data->pending_event.point.y;
+	} else if (cap->current_orientation == DISPLAY_ORIENTATION_ROTATED_270) {
+		lv_coord_t x;
```
same

fabiobaltieri commented on modules/lvgl/input/lvgl_pointer_input.c:
```diff
@@ -0,0 +1,133 @@
+	}
+}
+
+int lvgl_pointer_input_init(const struct device *dev)
```
static

fabiobaltieri commented on modules/lvgl/input/lvgl_pointer_input.c:
```diff
@@ -0,0 +1,133 @@
+#define LVGL_POINTER_INPUT_DEFINE(_inst)                                                           \
+	LVGL_INPUT_DECLARE(_inst, pointer, CONFIG_LV_Z_POINTER_INPUT_MSGQ_COUNT);                  \
+	static const struct lvgl_pointer_input_config lvgl_pointer_input_config##_inst = {         \
+		.common_config.input_dev = LVGL_INPUT_DEVICE(_inst),                               \
```
Is this used?

faxe1008 replied:
It was not, good catch :^)

fabiobaltieri commented on samples/subsys/display/lvgl/boards/stm32h7b3i_dk.overlay:
```diff
@@ -0,0 +1,16 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
drop one blank line, also the ones on line 9 and 15, keep it compact, same in the other overlay files

fabiobaltieri reviewed (commented):
Cool, did a first round of comments. Thinking you may want to move commit 2 and 3 in their own PR so they are out of the way since they are loosely coupled with this one and may as well just get merged.

fabiobaltieri commented on modules/lvgl/Kconfig.input:
```diff
@@ -38,4 +42,19 @@ config LV_Z_POINTER_KSCAN_INVERT_Y
+	  Size of the pointer message queue buffering input events.
+
+
+
```
nit: bunch of empty lines?  just keep one

fabiobaltieri commented on modules/lvgl/Kconfig.input:
```diff
@@ -38,4 +42,19 @@ config LV_Z_POINTER_KSCAN_INVERT_Y
+	depends on DT_HAS_ZEPHYR_LVGL_POINTER_INPUT_ENABLED
+
+config LV_Z_POINTER_INPUT_MSGQ_COUNT
+	int "Message queue size"
```
this (and the button one on the last patch) should be something like "Input pointer queue message count" (better qualified, and less ambigious, size feels like bytes when it's actually messages)

fabiobaltieri commented on modules/lvgl/input/lvgl_button_input.c:
```diff
@@ -0,0 +1,97 @@
+	data->pending_event.state = evt->value ? LV_INDEV_STATE_PR : LV_INDEV_STATE_REL;
+
+	if (k_msgq_put(cfg->common_config.event_msgq, &data->pending_event, K_NO_WAIT) != 0) {
+		LOG_DBG("Could not put input data into queue");
```
should probably be a LOG_WRN

fabiobaltieri commented on modules/lvgl/input/lvgl_button_input.c:
```diff
@@ -0,0 +1,97 @@
+	BUILD_ASSERT(LVGL_KEY_VALID(DT_INST_PROP(inst, lvgl_key)),                                 \
+		     "Property lvgl-key needs to be in range [0,255]")
+
+#define LVGL_BUTTON_INPUT_DEFINE(_inst)                                                            \
```
`inst`

fabiobaltieri commented on dts/bindings/input/zephyr,lvgl-button-input.yaml:
```diff
@@ -0,0 +1,47 @@
+          y-coordinate = <120>;
+  };
+
+compatible: "zephyr,lvgl-button-input"
```
good stuff! you should add something that builds this in CI, the lvgl sample already takes the button to reset the counter, how about moving that code to use input as well (just let it hit `INPUT_KEY_0`, most boards have it) and then add a config on this one? May even be a separate PR if you fancy it.

faxe1008 replied:
Just checked, adding  general purpose might be a bit tricky, since I at the moment the coordinates for the button are hardcoded on the device tree overlay. It's possible to assign them at runtime, but I would need to reach into the `lvgl_common_input_data` struct.  
Since the it is assumed that `lvgl_common_input_data` is the first member of any pseudo device `data` struct this would be doable, but I also adds that assumption to the sample. Otherwise a dedicated header may be added, which encapsulates the call. Something like:

```
lv_indev_t* lvgl_input_get_indev(const struct* device);
```

This will eventually be need anyways I guess, since the rotary encoders `indev` wants to be assigned a `lv_group_t` to function properly.

fabiobaltieri replied:
Oh I was thinking about just adding an example fixed instance as part of the sample, just to have the code build, maybe pointing to an LVGL keycode instead coordinate so it does not depend on the screen size? Alternatively you could start new `tests/drivers/build_all`.

faxe1008 replied:
👍 I added an `lvgl-button-input` to the `native_posix` overlay. This will trigger a click on the "hello world" button in the sample when pressing 'b' on the keyboard (I extended the sdl_gpio).

fabiobaltieri commented on modules/lvgl/input/lvgl_common_input.h:
```diff
@@ -0,0 +1,50 @@
+#define LVGL_INPUT_DEVICE(inst)           DEVICE_DT_GET_OR_NULL(DT_INST_PHANDLE(inst, input))
+
+#define LVGL_COORD_VALID(coord) (coord >= LV_COORD_MIN && coord <= LV_COORD_MAX)
+#define LVGL_KEY_VALID(key)     (key >= 0 && key <= 255)
```
this could be `IN_RANGE(key, 0, UINT8_MAX)`, up to you really

faxe1008 committed:
samples: subsys: display: lvgl: add lvgl-button-input to native_posix

Adds a lvgl-button-input pseudo device to the lvgl sample. Triggering the
SDL gpio (pressing 'b') on the keyboard should trigger a click event at the
center of the screen.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

gmarull commented on modules/lvgl/input/lvgl_common_input.h:
```diff
@@ -0,0 +1,50 @@
+#define LVGL_COORD_VALID(coord) IN_RANGE(coord, LV_COORD_MIN, LV_COORD_MAX)
+#define LVGL_KEY_VALID(key)     IN_RANGE(key, 0, UINT8_MAX)
+
+#define LVGL_INPUT_DECLARE(inst, type, msgq_size, process_evt_cb)                                  \
```
nit: this is not declaring but defining

gmarull commented on dts/bindings/input/zephyr,lvgl-pointer-input.yaml:
```diff
@@ -0,0 +1,36 @@
+  Needs to be associated to a specific device to listen for events
+  from that device. Example configuration:
+
+  pointer {
+          compatible = "zephyr,lvgl-pointer-input";
+          input = <&input_sdl_touch>;
+  };
```
shouldn't this be a child of `input_sdl_touch`?

fabiobaltieri replied:
Making it a child device means you can no longer instantiate the driver unattached from any device (that is, no `input` phandles specified, the callback receive any input event). I'm thinking that that may be fine for touchscreen, but probably not for other type of input, for example input receiving input codes from multiple devices, so we may as well keep them coherent and have them all with an `input` property. Not a strong opinion though, what do you think?

faxe1008 replied:
I think for the pointer its fine either way, for the buttons it makes sense to have it flexible. The only real upside for leaving it as is is the consistency between the pseudo-devices.

gmarull commented on modules/lvgl/input/lvgl_button_input.c:
```diff
@@ -0,0 +1,96 @@
+
+#define DT_DRV_COMPAT zephyr_lvgl_button_input
+
+#include "lvgl_common_input.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(lvgl);
```
```suggestion
#include "lvgl_common_input.h"

#include <zephyr/logging/log.h>

LOG_MODULE_DECLARE(lvgl);
```

gmarull commented on modules/lvgl/input/lvgl_common_input.c:
```diff
@@ -0,0 +1,47 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/device.h>
+#include <zephyr/input/input.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/display.h>
+
+#include "lvgl_common_input.h"
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(lvgl);
```
```suggestion
#include "lvgl_common_input.h"

#include <zephyr/device.h>
#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>

LOG_MODULE_DECLARE(lvgl);
```

please make sure to sort headers properly (local, zephyr, 3rd party) and do not include stuff that is not needed.

gmarull commented on modules/lvgl/input/lvgl_common_input.h:
```diff
@@ -0,0 +1,50 @@
+#ifndef ZEPHYR_MODULES_LVGL_LVGL_COMMON_INPUT_H_
+#define ZEPHYR_MODULES_LVGL_LVGL_COMMON_INPUT_H_
+
+#include <lvgl.h>
+#include <zephyr/device.h>
+#include <zephyr/input/input.h>
```
```suggestion
#include <zephyr/device.h>
#include <zephyr/kernel.h>
#include <zephyr/sys/util.h>

#include <lvgl.h>
```

same here

gmarull commented on modules/lvgl/input/lvgl_pointer_input.c:
```diff
@@ -0,0 +1,130 @@
+
+#define DT_DRV_COMPAT zephyr_lvgl_pointer_input
+
+#include "lvgl_common_input.h"
+#include <zephyr/drivers/display.h>
+#include <lvgl_display.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(lvgl);
```
same

fabiobaltieri commented on dts/bindings/input/zephyr,lvgl-button-input.yaml:
```diff
@@ -0,0 +1,47 @@
+          compatible = "zephyr,lvgl-button-input";
+          input = <&buttons>;
+          input-codes = <INPUT_KEY_0>;
+          lvgl-key = <10>;
+          x-coordinate = <120>;
+          y-coordinate = <120>;
```
hey, sorry caught this one late, trying to understand the LVGL button device a bit better. Would it make sense to have these three as arrays so that `input-codes` map to pairs of `x-coordinates` `y-coordinates`? Then about `lvgl-key`, does that really have to be an `LV_KEY_*`? From what I understand that should instead be the index of the x,y pair passed `points_array` on device registration.

So I'm thinking this should be specified as, for example:
```
          input-codes = <INPUT_KEY_0, INPUT_KEY_0>;
          x-coordinate = <120, 110>;
          y-coordinate = <120, 110>;
```

Where the pairs `120, 120` and `110, 110` would be registered and `lvgl_button_process_event` would just send, say key offset 0, 1 etc...

faxe1008 replied:
Ahhhhh, right. I thoroughly misunderstood that one. I thought that the logic of the lvgl indev was that a key down event with the `LV_KEY_*` at all of the locations was raised. That was why I intially had only support for one point added because I did not see why a `multiclick` would make sense. Rather the `btn_id` is basically indexing into the the coordinates array. This also has the benefit that you really only need one pseudo device for X physical buttons. Neat.

Thanks a lot for raising this  :heart: . Maybe I'll open a PR upstream and a short paragraph explaining this :^)

fabiobaltieri replied:
Yeah I mixed them up as well, I find the doc unclear, ended up checking the code directly: https://github.com/zephyrproject-rtos/lvgl/blob/8a6a2d1d29d17d1e4bdc94c243c146a39d635fdd/src/core/lv_indev.c#L768

faxe1008 replied:
Added the implementation according to your suggestion, works like a charm :^)

faxe1008 replied:
Raised a smol PR upstream to update docs :^):
[lvgl/pull/4478](https://github.com/lvgl/lvgl/pull/4478)

fabiobaltieri commented on modules/lvgl/input/lvgl_button_input.c:
```diff
@@ -0,0 +1,96 @@
+		return;
+	}
+
+	data->pending_event.key = cfg->key;
```
this should be `btn_id` for `LV_INDEV_TYPE_BUTTON`

faxe1008 commented on dts/bindings/input/zephyr,lvgl-button-input.yaml:
```diff
@@ -0,0 +1,36 @@
+          compatible = "zephyr,lvgl-button-input";
+          input = <&buttons>;
+          input-codes = <INPUT_KEY_0 INPUT_KEY_1>;
+          coordinates = <120 220 150 250>;
```
Did not know that this works. Neat :+1:

gmarull commented:
ping @jfischer-no

jfischer-no reviewed (approved):
LGTM

GeorgeCGV commented:
There is a minor issue with the input pointer with lvgl.
When `CONFIG_INPUT_THREAD_PRIORITY` is lower than that of a thread that runs `lv_task_handler`, the lvgl might process the same event more than once.

GeorgeCGV commented:
@fabiobaltieri investigated it a bit more. The issue turned out in a custom LVGL component. Interesting that it showed up only after switching to `lvgl-input-pointer`. Sorry for the inconvenience.

fabiobaltieri commented:
> @fabiobaltieri investigated it a bit more. The issue turned out in a custom LVGL component. Interesting that it showed up only after switching to `lvgl-input-pointer`. Sorry for the inconvenience.

Oh ok, maybe something changed from the previous implementation, anyway thanks for the update.
